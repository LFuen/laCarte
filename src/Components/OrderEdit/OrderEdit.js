import React, {Component} from 'react'
import LaCarteContext from '../../context/LaCarteContext'
import PropTypes from 'prop-types'
import api from '../../config'
import '../OrderForm/OrderForm.css'


class OrderEdit extends Component{

    static propTypes = {
        match: PropTypes.shape({
            params: PropTypes.object
        }),
        history: PropTypes.object
    }
    static contextType = LaCarteContext

    state = {
        error: null,
    }

    handleSubmit = event => {
        event.preventDefault()

        const idParam = this.props.match.params.id

        const {houseadd, apartment, city, state, zip, tel, meal} = event.target
        
        const order = {
            prim_add: houseadd.value,
            sec_add: apartment.value,
            city: city.value,
            state: state.value,
            zip: zip.value,
            phone: tel.value,
            meal: meal.value
        }
        this.setState({error: null})
        fetch(`${api.API_ENDPOINT}/api/orders/${idParam}`, {
            method: 'PATCH',
            body: JSON.stringify(order),
            headers: {
                'content-type': 'application/json',
                'Authorization': `Bearer ${api.API_KEY}`
            }
        })
        .then(res => {
            if(!res.ok) {
                return res.json().then(error => {
                    throw error
                })
            }
        })
        .then(() => {
            order.id = Number(idParam)
            this.context.updateOrder(order)
            this.props.history.push('/OrderUpdated')
        })
        .catch(error => {
            this.setState({error})
        })
    }


    render(){

        const {orders} = this.context

        if(orders.length > 0){
            const idParam = this.props.match.params.id
            const order = orders.find(order => order.id === Number(idParam));

            return(
            <div className='container'>
                <div className='title'>
                    <h2>Change Order</h2>
                    <div className='d-flex'>
                        <form onSubmit={this.handleSubmit} method='' className='orderForm'>
                            <label>
                            <span>Meal</span><br/>
                            <input id='meal' type='text' name='meal' value={order.meal} required readOnly/>
                            </label>
                            <label>
                            <span>Street Address *</span><br/>
                            <input id='houseadd' type='text' name='houseadd' placeholder='House number and street name' required/>
                            </label>
                            <label>
                            <span>Address 2</span><br/>
                            <input id='apartment' type='text' name='apartment' placeholder='Apartment, suite, unit etc. (optional)'/>
                            </label>
                            <label>
                            <span>City *</span><br/>
                            <input id='city' type='text' name='city' placeholder='City'/> 
                            </label>
                            <label>
                            <span>State / County *</span><br/>
                            <input id='state' type='text' name='state' placeholder='State'/> 
                            </label>
                            <label>
                            <span>Postcode / ZIP *</span><br/>
                            <input id='zip' type='text' name='zip' pattern='[0-9]{5}' placeholder='Zip Code'/> 
                            </label>
                            <label>
                            <span>Phone *</span><br/>
                            <input id='tel' type='tel' name='tel' pattern='[0-9]{3}-[0-9]{3}-[0-9]{4}' placeholder='000-000-000'/>
                            </label>
                            <button type='submit' className='shadow'>
                                Save Order
                            </button>
                        </form>
                    </div>
                </div>
            </div>
        )} else {
            return (<h2>Loading..</h2>)
        }
    }
}


export default OrderEdit