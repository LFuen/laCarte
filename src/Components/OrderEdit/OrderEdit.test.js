import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'
import {BrowserRouter} from 'react-router-dom'
import LaCarteContext from '../../context/LaCarteContext'
import OrderEdit from './OrderEdit'



describe(`OrderEdit Component`, () => {
    describe(`Snapshot`, () => {
        it(`renders the UI as expected`, () => {
            const edit = renderer.create(<BrowserRouter><OrderEdit/></BrowserRouter>)
                expect(edit).toMatchSnapshot()
        })
    })

    describe(`Smoke test`, () => {
        it(`Renders without crashing`, () => {
            
            const div = document.createElement('div');
            const renderer = new ShallowRenderer()

            const value = {
                meals: 'string',
                chefs: 'string',
                cuisines: 'string',
                orders: 'string',
                addOrder: 'string',
                updateOrder: 'string',
                deleteOrder: 'string',
            };

            renderer.render(
                <BrowserRouter>
                    <LaCarteContext.Provider value={value}>
                        <OrderEdit />
                    </LaCarteContext.Provider>                
                </BrowserRouter>,
                div)
            ReactDOM.unmountComponentAtNode(div)
        })
    })
})