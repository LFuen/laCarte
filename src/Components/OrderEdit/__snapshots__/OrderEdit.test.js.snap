// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderEdit Component Snapshot renders the UI as expected 1`] = `
<h2>
  Loading..
</h2>
`;
